File: packages/python/octospaces-sdk/octospaces_sdk/utils/invite_preview.py
"""Invite preview — classify an invite string without joining.

Mirrors ``packages/ts/octospaces-sdk/src/utils/invite-preview.ts``.

Three recognised kinds:
- ``space-link``   — base64url-encoded JSON token in the URL fragment (``/join#<b64>``)
- ``node-link``    — same but ``/join/node#<b64>``
- ``member-bundle`` — plain JSON with ``spaceId``, ``cap``, and ``inviteJson``
"""

from __future__ import annotations

import json
from typing import Any, Literal, Optional, TypedDict


class SpaceLinkPreview(TypedDict):
    kind: Literal["space-link"]
    space_name: str
    write: bool
    token: dict[str, Any]


class NodeLinkPreview(TypedDict):
    kind: Literal["node-link"]
    space_name: str
    node_title: Optional[str]
    token: dict[str, Any]


class MemberBundlePreview(TypedDict):
    kind: Literal["member-bundle"]
    space_name: str
    space_id: str
    issuer_key: Optional[str]
    invite_json: str


InvitePreview = SpaceLinkPreview | NodeLinkPreview | MemberBundlePreview


def _decode_b64url_json(fragment: str) -> dict[str, Any]:
    import base64

    b64 = fragment
    # Restore padding
    padded = b64 + "=" * ((4 - len(b64) % 4) % 4)
    raw = base64.urlsafe_b64decode(padded).decode()
    return json.loads(raw)


def preview_invite(raw: str) -> InvitePreview:
    """Classify *raw* as a space-link, node-link, or member-bundle.

    Raises :exc:`ValueError` if *raw* is not recognisable.
    """
    raw = raw.strip()

    # Check for URL fragment link patterns
    node_marker = "/join/node#"
    space_marker = "/join#"

    if node_marker in raw:
        fragment = raw.split(node_marker, 1)[1]
        token = _decode_b64url_json(fragment)
        return NodeLinkPreview(
            kind="node-link",
            space_name=token.get("spaceName", ""),
            node_title=token.get("nodeName"),
            token=token,
        )

    if space_marker in raw:
        fragment = raw.split(space_marker, 1)[1]
        token = _decode_b64url_json(fragment)
        return SpaceLinkPreview(
            kind="space-link",
            space_name=token.get("spaceName", ""),
            write=bool(token.get("write", False)),
            token=token,
        )

    # Try as JSON member-bundle
    try:
        bundle = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise ValueError(f"Unrecognised invite format: {raw!r}") from exc

    space_id: str = bundle.get("spaceId", "")
    space_name: str = bundle.get("spaceName", "")
    invite_json: str = raw
    # Issuer key: the cap's ``iss`` field if present
    cap = bundle.get("cap") or {}
    issuer_key: Optional[str] = cap.get("iss") if isinstance(cap, dict) else None

    return MemberBundlePreview(
        kind="member-bundle",
        space_name=space_name,
        space_id=space_id,
        issuer_key=issuer_key,
        invite_json=invite_json,
    )

File: packages/ts/octospaces-sdk/src/utils/invite-preview.ts
/**
 * Parse an invite (pasted text or a `#fragment` deep link) into a preview the
 * join screen can show on a consent card — name, type, identifying fingerprint
 * — WITHOUT joining. Actual join calls (`joinSpaceByLink`, `joinNodeByLink`,
 * `acceptSpaceInvite`, `acceptNodeInvite`) only run after the user confirms.
 *
 * Accepts three input forms:
 *   - A space invite link (URL fragment encoded by `createSpaceInviteLink`)
 *   - A node invite link (URL fragment encoded by `createNodeInviteLink`)
 *   - A private member-bundle JSON minted by `inviteToSpace`
 */
import { decodeSpaceInviteLink, decodeNodeInviteLink } from '@drakkar.software/starfish-spaces';
import type { SpaceInviteLinkToken, NodeInviteLinkToken } from '@drakkar.software/starfish-spaces';

export type { SpaceInviteLinkToken, NodeInviteLinkToken };

export type InvitePreview =
  | {
      kind: 'space-link';
      spaceName: string;
      /** True if the link grants write access, false for read-only. */
      write: boolean;
      token: SpaceInviteLinkToken;
    }
  | {
      kind: 'node-link';
      spaceName: string;
      /** The node's display name, absent for legacy tokens that omit it. */
      nodeTitle?: string;
      token: NodeInviteLinkToken;
    }
  | {
      kind: 'member-bundle';
      spaceName: string;
      spaceId: string;
      /** Short hex fingerprint of the issuing owner's signing key, or null if absent. */
      issuerKey: string | null;
      /** The raw cap-bundle JSON — pass verbatim to `acceptSpaceInvite` on consent. */
      inviteJson: string;
    };

/** Shape of the private invite bundle minted by `inviteToSpace`. */
interface PrivateInviteShape {
  spaceId?: string;
  spaceName?: string;
  cap?: { kind?: string; iss?: string };
}

/**
 * Classify and decode an invite string into a typed {@link InvitePreview}.
 * Throws a human-readable `Error` on invalid input (safe to surface verbatim
 * in a toast or inline error message).
 */
export function previewInvite(raw: string): InvitePreview {
  const text = raw.trim();
  if (!text) throw new Error('Paste an invite link or code first.');

  // Invite links carry their credential in a URL fragment.
  if (text.includes('#')) {
    const fragment = text.slice(text.indexOf('#'));
    // Try node-invite link first — it has a `nodeId` field the space link lacks.
    try {
      const token = decodeNodeInviteLink(fragment);
      return {
        kind: 'node-link',
        spaceName: `space-${token.spaceId.slice(-6)}`,
        nodeTitle: token.nodeName,
        token,
      };
    } catch {
      // fall through to space link
    }
    try {
      const token = decodeSpaceInviteLink(fragment);
      return { kind: 'space-link', spaceName: token.spaceName, write: token.write, token };
    } catch {
      throw new Error('That invite link appears to be invalid or expired.');
    }
  }

  // Private member-bundle JSON minted by `inviteToSpace`.
  let parsed: PrivateInviteShape;
  try {
    parsed = JSON.parse(text) as PrivateInviteShape;
  } catch {
    throw new Error("That doesn't look like an invite. Paste the full invite code or link.");
  }
  if (!parsed?.spaceId || parsed.cap?.kind !== 'member') {
    throw new Error('That is not a valid space invite.');
  }
  const iss = parsed.cap?.iss;
  return {
    kind: 'member-bundle',
    spaceName: parsed.spaceName?.trim() || `space-${parsed.spaceId.slice(-6)}`,
    spaceId: parsed.spaceId,
    issuerKey: typeof iss === 'string' && iss.length >= 8 ? `${iss.slice(0, 8)}…${iss.slice(-8)}` : null,
    inviteJson: text,
  };
}
